### ISSUE TITLE
Progressbar is reset when window is inactive

### ISSUE THREAD

**Issue by elmoro666** [needim/noty#357]
> **_Noty version_**: 3.1.0
> **_Browser_**: Chrome
> **_OS_**: Windows 10
> **_jsFiddle link (if available)_**: .....
----------

There has been a change in behavior in the beta versions compared to the old. Let's say you have a noty warning with timeout of 7 seconds and a progress bar. While the progress bar is running, go to another tab in chrome for a few seconds, and then come back. You will see that the progress bar has been reset. This did not happen in the old version.  I have added a simple demo to demonstrate this. Just unzip and click on index.html.

[Noty.zip](https://github.com/needim/noty/files/952509/Noty.zip)

**Comment by needim** [comment]
Yes, that behaviour is intentional, to ensure that users do not miss their notifications

**Comment by elmoro666** [comment]
Ok thanks. Is it a lot of work to add an option to get the old behaviour?

The problem is that the new behavior ensures that we do **not** get the notifications. We use SignalR/Websockets to alert when something happens. We all have three monitors, and if I happen to work in another application while the website is on the other screen, I do not see the notifications. When I activate the page again I get all the notifications, but that is too late.

**Comment by needim** [comment]
Yeah, It will be nice to have an option for that behaviour. Something like "**visibilityControl**"
I will implement that option for next release.

**Comment by elmoro666** [comment]
awesome! Thanks for your efforts

**Comment by needim** [comment]
visibilityControl option added. I will update documentation soon.

**Comment by KrunchMuffin** [comment]
Just ran into this. is it just boolean?

**Comment by needim** [comment]
@KrunchMuffin yes, it's boolean and default is **true**.  Set **false** for disable this behaviour.

**Comment by KrunchMuffin** [comment]
I am not quite sure what is going on, but the `visibilityControl: false` is not working in all instances.

**Comment by needim** [comment]
Are you using overrideDefaults method or passing as a option parameter?

**Comment by KrunchMuffin** [comment]
overrideDefaults  method

**Comment by KrunchMuffin** [comment]
`Noty.overrideDefaults({
  layout: 'topRight',
  theme: 'sunset',
  visibilityControl: false,
  timeout: '7000'
})
`

**Comment by needim** [comment]
Wow, that's weird. Can you provide more detail about the case? Maybe we can reproduce the issue and catch the error.

**Comment by KrunchMuffin** [comment]
I'll see what I can come up with

**Comment by KrunchMuffin** [comment]
I commented out  
103 https://github.com/needim/noty/blob/master/lib/noty.js#L103
2830 https://github.com/needim/noty/blob/master/lib/noty.js#L2830

and it worked as expected.

**Comment by KrunchMuffin** [comment]
I uncommented and commented out

https://github.com/needim/noty/blob/master/lib/noty.js#L357

and it worked as expected.

**Comment by needim** [comment]
Yeah, but this way you removed entire feature.

Some questions;
- Are you calling overrideDefaults (just once) before all the Noty's right?
- Are you using requirejs or some kind of umd/amd loader? Are you sure about noty.js loaded just once?
- Are you using Noty with Vue, Angular, React, e.g? (just to know about environment, I don't think it's related with issue)
- Did you tried give as an option for all Noty's without overriding?

**Comment by KrunchMuffin** [comment]
Ya, I know, but I am trying to narrow down.

Yes
Using LABjs
No
Yes

so some pseudo code
PageHidden = true
if hidden call stopAll()

inside stopAll the only console.log that is output is first one `in stopall`

`  function stopAll() {
    console.log("in stopall")
    setTimeout(function () {
      Object.keys(API.Store).forEach(function (id) {
        console.log(id)
        if (API.Store.hasOwnProperty(id)) {
            console.log("hasOwnProperty true")
          if (API.Store[id].options.visibilityControl) {
            console.log(API.Store[id].options.visibilityControl)
            API.Store[id].stop();
          }
        } else {
            console.log("hasOwnProperty false")
        }
      });
    }, 100);
  }`

**Comment by KrunchMuffin** [comment]
`Object.keys(API.Store)` has 0 length

**Comment by needim** [comment]
If store is empty there is no active notification on the screen.

Can you overrideDefault callback for debug. Lets watch console for booleans.

```javascript 
callbacks: {
    beforeShow: function() {
    	console.log(this.options.visibilityControl)
    }
}
```

**Comment by KrunchMuffin** [comment]
ya, that is true because I have not give it focus yet. Shouldn't it ignore that and just continue to show the alert anyway?

**Comment by KrunchMuffin** [comment]
ok, so I put the callback in. When the window is blurred, nothing happens, no log. When I focus it, it shows and comes back `false`

**Comment by needim** [comment]
Oh right! Now I see the case. There is a missing implementation in Noty's show function. We should check visibilityControl there too. NICE CATCH!

https://github.com/needim/noty/blob/master/src/index.js#L77

It will be fixed in next release. Sorry about that. :/ Thanks.

For quick fix you can commentout or remove line 2830
https://github.com/needim/noty/blob/master/lib/noty.js#L2830

**Comment by KrunchMuffin** [comment]
Cool, thx for the help.

**Comment by needim** [comment]
No, thanks for you! 👍

**Comment by KrunchMuffin** [comment]
Can you also add a header at the top of the files that shows what version it is and any other info like a URL to ur site?

**Comment by needim** [comment]
Oh, sure nice point. I will add a banner for next releases. 
(currently there is the only way to get the version is calling Noty.version())

**Comment by nordicbondpricing** [comment]
One more thing related to this (I can open a new issue if you prefer):

If the browser (with noty alert js code) is not active, the alerts won't display. Imagine you have a sports website with alerts showing scores in different matches. If the user is doing something else on one his/her other monitors, the alerts will not display. If you reactivate the page, the alerts display. This is how it worked pre-beta by the way. I think the old behaviour was better, but it could also be an option. What do you think?

**Comment by elmoro666** [comment]
One more thing related to this (I can open a new issue if you prefer):

If the browser (with noty alert js code) is not active, the alerts won't display. Imagine you have a sports website with alerts showing scores in different matches. If the user is doing something else on one his/her other monitors, the alerts will not display. If you reactivate the page, the alerts display. This is how it worked pre-beta by the way. I think the old behaviour was better, but it could also be an option. What do you think?

**Comment by KrunchMuffin** [comment]
That is pretty much the whole point of this thread, hence why the new option in place.

**Comment by needim** [comment]
@elmoro666 yep, we were talking about this behaviour. No need a new issue.

**Comment by needim** [comment]
Fixed in v3.1.1, sorry for the late release
@KrunchMuffin also banner added in this version.
